src/test/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitCheckerTest.java

```diff
@@ -0,0 +1,301 @@
+
+        handler = new RingBufferLogHandler(100);
+        final Logger logger = Logger.getLogger(getClass().getName());
+        logger.addHandler(handler);
+        listener = new LogTaskListener(logger, Level.INFO);
+    }
+
+    private void setupStubs(List<JsonRateLimit> scenarios) {
+        githubApi.listAllStubMappings().getMappings().stream()
+                .filter(x -> x.getScenarioName() != null && x.getScenarioName() == ("API Rate Limit"))
+                .forEach(x -> githubApi.removeStubMapping(x));
+
+        for (int i = 0; i < scenarios.size(); i++) {
+            String state = (i == 0) ? Scenario.STARTED : Integer.toString(i);
+            String nextState = Integer.toString(i + 1);
+            JsonRateLimit scenarioResponse = scenarios.get(i);
+
+            String limit = Integer.toString(scenarioResponse.rate.limit);
+            String remaining = Integer.toString(scenarioResponse.rate.remaining);
+            String reset = Long.toString(scenarioResponse.rate.reset.toInstant().getEpochSecond());
+            String body = "{ \"rate\": { \"limit\": " + limit +  ", " +
+                    "\"remaining\": " + remaining + ", " +
+                    "\"reset\": " + reset + " } }";
+            ScenarioMappingBuilder scenario = get(urlEqualTo("/rate_limit"))
+                    .inScenario("API Rate Limit")
+                    .whenScenarioStateIs(state)
+                    .willReturn(aResponse()
+                            .withHeader("Content-Type", "application/json; charset=utf-8")
+                            .withHeader("X-RateLimit-Limit", limit)
+                            .withHeader("X-RateLimit-Remaining", remaining)
+                            .withHeader("X-RateLimit-Reset", reset)
+                            .withBody(body)
+                    );
+            if (i != scenarios.size() - 1) {
+                scenario = scenario.willSetStateTo(nextState);
+            }
+            githubApi.stubFor(scenario);
+
+        }
+    }
+
+    private Date soon = Date.from(LocalDateTime.now().plusMinutes(60).atZone(ZoneId.systemDefault()).toInstant());
+
+    @Test
+    public void ThrottleOnOverTestWithQuota() throws Exception {
+        // set up checker: don't actually sleep
+        ApiRateLimitChecker.setEntropy(entropy);
+        ApiRateLimitChecker.setExpirationWaitMillis(1);
+        ApiRateLimitChecker.setNotificationWaitMillis(1);
+
+        // set up scenarios
+        List<JsonRateLimit> scenarios = new ArrayList<>();
+        int limit = 5000;
+        scenarios.add(new JsonRateLimit(limit, limit, soon));
+        setupStubs(scenarios);
+
+        // Given a full rate limit quota, then we expect no throttling
+        for (int i = 0; i < 100; i++) {
+            apiRateLimitCheckerThrottleOnOver.checkApiRateLimit(listener, github);
+        }
+
+        assertEquals(0, handler.getView().stream().map(LogRecord::getMessage).filter(m -> m.contains("Sleeping")).count());
```

Comment:
Turn this in to one more helper methods.   

The simplest thing to do would be to have something that returns and `List<String>` or `Iterable<String`.  Then you could use standard hamcrest matchers to `assertThat` the list does or does not contain the expected outputs.  

Alternatively, you could make higher level helpers such as `assertSleepCount(expectedNumThrottles)` and so on. 

But for more complex scenarios, you might need to check the order and count of messages.  Shrug.